src/audio/pipeline.c

```diff
@@ -844,20 +870,22 @@ void pipeline_xrun(struct pipeline *p, struct comp_dev *dev,
 	if (dev->state != COMP_STATE_ACTIVE)
 		return;
 
+	memset(&posn, 0, sizeof(posn));
+	p->xrun_bytes = bytes;
+	posn.xrun_size = bytes;
+	posn.xrun_comp_id = dev->comp.id;
+	data.posn = &posn;
+
 	/* notify all pipeline comps we are in XRUN, and stop copying */
+	irq_local_disable(flags);
 	ret = pipeline_trigger(p, p->source_comp, COMP_TRIGGER_XRUN);
 	if (ret < 0)
 		trace_pipe_error_with_ids(p, "pipeline_xrun() error: "
 					  "Pipelines notification about XRUN "
 					  "failed, ret = %d", ret);
 
-	memset(&posn, 0, sizeof(posn));
-	p->xrun_bytes = bytes;
-	posn.xrun_size = bytes;
-	posn.xrun_comp_id = dev->comp.id;
-	data.posn = &posn;
-
```

Comment:
Is it reasonable to do this move in a separate patch? (setting xrun_bytes before calling pipeline_trigger)

Reply:
@paulstelian97 don't see a reason. I'm adding an IRQ lock and moving code, that doesn't need to be protected out of the atomic context, it's perfectly related and otherwise has hardly value.

Reply:
Alright I guess.